Current task refactor

furquan-rivos opened the pull request.

furquan-rivos committed:
sys/kern: Move `switch_to` to task.rs

This change moves `switch_to` to task.rs so that it can be used by
more than just syscalls. Also, this function is more related to tasks
than syscalls. This change is being done to make `switch_to` as the
common entrypoint for any kernel path to set next user task that
should be run. This helps avoid duplication of some common work that
needs to be done irrespective of architecture.

furquan-rivos committed:
sys/kern: Use `switch_to` in arch-specific code

This change replaces all calls to {`apply_memory_protection` followed
by `set_current_task`} with `switch_to` since that is exactly what
`switch_to` does.

furquan-rivos committed:
sys/kern: Use `switch_to` in `start_first_task`

This change uses the `switch_to` function in the startup path as well
since we were doing the same thing there w.r.t. applying memory
protection and setting current task pointer. One difference in
behavior with this change would be that `apply_memory_protection` is
called just before jumping to the first task which should be okay.

TEST=Verified that rv32 tests still pass.

furquan-rivos committed:
sys/kern: Move call to `event_context_switch` to `switch_to`

This change moves the call to `crate::profiling::event_context_switch`
to `switch_to` instead of duplicating this in every architecture.

adurbin-rivos commented on sys/kern/src/arch/riscv32.rs:
```diff
@@ -607,11 +604,9 @@ unsafe fn handle_fault(task: *mut task::Task, fault: FaultInfo) {
-            // we're not going to read it back until the next kernel entry so
-            // we won't be aliasing/racing.
-            set_current_task(next);
+            // Safety: next comes from teh task table and we don't use it again
```
Not sure how much you care about fixing typos. 'teh' is in here in a few places from the existing code. however, I suspect the diff being generated more prominently shows it as part of the diff proper.

furquan-rivos replied:
Good point. Let me fix that in a follow-up commit.

furquan-rivos replied:
Done. Pushed a follow up change.

adurbin-rivos commented on sys/kern/src/startup.rs:
```diff
@@ -118,7 +118,6 @@ pub unsafe fn start_kernel(tick_divisor: u32) -> ! {
-    crate::arch::apply_memory_protection(&task_table[first_task_index]);
     TASK_TABLE_IN_USE.store(false, Ordering::Release);
     crate::arch::start_first_task(
         tick_divisor,
```
I would have like to put this one line below...

It's interesting that start_first_task was already being provided a mutable reference, but the compiler didn't care and dropped it. I'm more curious about the history here since it seems a mutable reference was necessary at some point.

Is the first task the idle task? Or is there a notion of this thread of context continuing past this line?

furquan-rivos replied:
You won't believe it, but as I was refactoring this, I got curious about the same thing and started playing with the code to see why the mutable reference is required. It looks like I can get rid of the mutable reference through the entire chain. I am still not sure why this mutable reference exists in the first place though. I will see if I can dig up some history.

> Is the first task the idle task? Or is there a notion of this thread of context continuing past this line?

First task could be anything that is marked with `START_AT_BOOT` flag in the app configuration: https://hubris.oxide.computer/reference/#_starting_the_first_tasks

This and `start_first_task` functions never return (return is defined as `!`).

FawazTirmizi replied:
I'm not sure about the mutable reference either. I was going to clean it up as a part of a patch making the kernel clippy clean.

adurbin-rivos replied:
> This and `start_first_task` functions never return (return is defined as `!`).

Thanks for the pointer. Good to know. Makes more sense then.

furquan-rivos replied:
[This](https://github.com/oxidecomputer/hubris/pull/582/commits/9cb4d9753c7a4c9f18fd993968dd49931a952c9e#diff-071eb4f76fe8d7a0d61bcc0454c6a2d03527130e5227e17f58dd758419469fef) seems to be the commit that introduced it. Still trying to make sense of it.

furquan-rivos replied:
Okay, so it looks like this is required for the [atomic store](https://doc.rust-lang.org/core/sync/atomic/struct.AtomicPtr.html#method.store)

This store method is called [here](https://github.com/oxidecomputer/hubris/blob/a6b0055273c067af60c0c76abde5a54c096f362e/sys/kern/src/arch/arm_m.rs#L655)

So, yeah I cannot get rid of the mutable reference.

adurbin-rivos replied:
I wasn't suggesting to get rid of it. I was noting it was already there prior. I guess we just didn't need it in the riscv port?

furquan-rivos replied:
Looks like our [RV port hasn't yet switched to using atomic ops](https://github.com/rivosinc/hubris/blob/rivos/main/sys/kern/src/arch/riscv32.rs#L706).

adurbin-rivos commented on sys/kern/src/arch/riscv32.rs:
```diff
@@ -429,10 +429,9 @@ fn timer_handler() {
                 // Safety: next comes from teh task table and we don't use it again
                 // until next kernel entry, so we meet the function requirements.
-                set_current_task(next);
+                crate::task::switch_to(next);
```
The naming is odd here, imo. switch_to() reads to me it's actually switching, but that's not what's happening. It's just setting up the bookkeeping for the eventual context switch proper. Maybe that's the semantics of switch_to(), but it seems odd to me.

adurbin-rivos replied:
"Performs the architecture-specific bookkeeping to activate `task` on next return to user." which is in the first commit's comments. I guess it's clearly just my opinion. :)

furquan-rivos replied:
Yeah, I hadn't changed the name, but I didn't like it too. Like you mentioned, my assumption on reading the function being used anywhere is that it is switching to the task, but clearly it does not. I thought of `select_next_task`, `setup_next_task`, `configure_next_switch`, `activate_next_task`. Not sure if either of these are any better though. If you have suggestion for a good name we can go with that.

furquan-rivos committed:
sys/kern: Fix typos

furquan-rivos committed:
sys/kern: Rename `switch_to` to `activate_next_task`

This change performs the following:
find . -type f -exec sed -i 's/switch_to/activate_next_task/g' {} \;

This change is done because the new name reflects the work done by the
function much better. `switch_to` could be confused by the reader to
mean that the function actually switches to the task.

furquan-rivos commented:
Renamed `switch_to` to `activate_next_task`.

furquan-rivos committed:
Run `cargo fmt`
